packages/database/src/useDatabaseSnapshot.ts

```diff
@@ -80,49 +55,29 @@ export type UseDatabaseValueOptions = {
 };
 
 export function useDatabaseValue<T = unknown | null, R = T>(
-  key: QueryKey,
+  queryKey: QueryKey,
   ref: DatabaseReference,
   options: UseDatabaseValueOptions = {},
   useQueryOptions?: Omit<UseQueryOptions<T, Error, R>, "queryFn">
 ): UseQueryResult<R, Error> {
-  const client = useQueryClient();
-  const unsubscribe = useRef<Unsubscribe>();
+  const isSubscription = !!options?.subscribe;
 
-  useEffect(() => {
-    return () => {
-      unsubscribe.current?.();
-    };
-  }, []);
-
-  return useQuery<T, Error, R>({
-    ...useQueryOptions,
-    queryKey: useQueryOptions?.queryKey ?? key,
-    staleTime:
-      useQueryOptions?.staleTime ?? options?.subscribe ? Infinity : undefined,
-    async queryFn() {
-      unsubscribe.current?.();
-      let resolved = false;
-
-      return new Promise<T>((resolve, reject) => {
-        unsubscribe.current = onValue(
-          ref,
-          (snapshot) => {
-            if (!resolved) {
-              resolved = true;
-              return resolve(parseDataSnapshot(snapshot, !!options?.toArray));
-            } else {
-              client.setQueryData<T>(
-                key,
-                parseDataSnapshot(snapshot, !!options?.toArray)
-              );
-            }
-          },
-          reject,
-          {
-            onlyOnce: !options.subscribe,
-          }
-        );
+  const subscribeFn = useCallback(
+    (callback: (data: T) => Promise<void>) => {
+      return onValue(ref, (snapshot) => {
+        const data = parseDataSnapshot(snapshot, !!options?.toArray);
+        return callback(data);
       });
     },
-  });
+    [ref]
+  );
+
+  return useSubscription<T, Error, R>(
+    queryKey,
+    ["useFirestoreDatabase", ref.key],
+    subscribeFn,
+    useQueryOptions,
+    !isSubscription,
+    async () => parseDataSnapshot(await get(ref), !!options?.toArray)
+  );
```

Comment:
```suggestion
return useSubscription<T, Error, R>(
    queryKey,
    ['useDatabaseValue', queryKey],
    subscribeFn,
    {
      ...useQueryOptions,
      onlyOnce: !isSubscription,
      fetchFn: async () => parseDataSnapshot(await get(ref), !!options?.toArray),
    },
  );
```
1. `useSubscription` has 4 args. The code update seems to be missing.
2. The reason for changing from 'ref.key' to 'queryKey' is that I found that a bug occurs when 'ref' is used as a query. When I changed it to 'queryKey', I checked that it works well.